packages/gatsby/src/schema/node-model.js

```diff
@@ -168,40 +202,137 @@ class LocalNodeModel {
       `Querying GraphQLUnion types is not supported.`
     )
 
-    // We provide nodes in case of abstract types, because `run-sift` should
-    // only need to know about node types in the store.
-    let nodes
+    const fields = getQueryFields({
+      filter: query.filter,
+      sort: query.sort,
+      group: query.group,
+      distinct: query.distinct,
+    })
+    const fieldsToResolve = determineResolvableFields(
+      this.schemaComposer,
+      this.schema,
+      gqlType,
+      fields
+    )
+
     let nodeTypeNames
     if (isAbstractType(gqlType)) {
       nodeTypeNames = toNodeTypeNames(this.schema, gqlType)
-      nodes = nodeTypeNames.reduce(
-        (acc, typeName) => acc.concat(this.nodeStore.getNodesByType(typeName)),
-        []
-      )
     } else {
       nodeTypeNames = [gqlType.name]
     }
 
+    await this.prepareNodes(gqlType, fields, fieldsToResolve, nodeTypeNames)
+
     const queryResult = await this.nodeStore.runQuery({
       queryArgs: query,
       firstOnly,
+      gqlSchema: this.schema,
+      gqlComposer: this.schemaComposer,
       gqlType,
-      nodes,
+      resolvedFields: fieldsToResolve,
       nodeTypeNames,
     })
 
     let result = queryResult
     if (args.firstOnly) {
       if (result && result.length > 0) {
         result = result[0]
+        this.trackInlineObjectsInRootNode(result)
       } else {
         result = null
       }
+    } else if (result) {
+      result.forEach(node => this.trackInlineObjectsInRootNode(node))
     }
 
     return this.trackPageDependencies(result, pageDependencies)
   }
 
+  prepareNodes(type, queryFields, fieldsToResolve, nodeTypeNames) {
+    const typeName = type.name
+    if (!this._prepareNodesQueues[typeName]) {
```

Comment:
Instead of running resolving immediately, we delay it.

This is because GraphQL is async, several resolvers that all need to prepare fields can query asynchronously. By queueing them, we make sure that all the promises inside this particular tick are queued, thus we shouldn't have conflicts.